### PR TITLE
Get rid of double exit/enter when we drag over hoverable component outside of window

### PR DESCRIPTION
**Pull request by igordmn**
Fixes https://github.com/JetBrains/compose-jb/issues/1218

Reproducer:
```
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.PointerEventType
import androidx.compose.ui.input.pointer.onPointerEvent
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.singleWindowApplication

@OptIn(ExperimentalComposeUiApi::class)
fun main() = singleWindowApplication {
    Box(
        Modifier
            .fillMaxSize()
            .padding(50.dp)
            .background(Color.Red)
            .onPointerEvent(PointerEventType.Exit) { println("EXIT") }
            .onPointerEvent(PointerEventType.Enter) { println("ENTER") }
    )
}
```

1. Press on the Red box
2. Move outside the box -> EXIT
3. Move outside the window -> EXIT
4. Move into the window -> ENTER  // hover is stuck here
5. Move into the box -> ENTER

Test: gradle compose:ui:ui:connectedCheck -Pandroidx.compose.multiplatformEnabled=false

### CODE REVIEW

File: compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/input/pointer/HitPathTracker.kt [hunk 1]
```diff
@@ -388,22 +388,21 @@ internal class Node(val pointerInputFilter: PointerInputFilter) : NodeParent() {
...
-                    )
+            if (event.type == PointerEventType.Move ||
+                event.type == PointerEventType.Enter ||
+                event.type == PointerEventType.Exit
             ) {
-                event.type = if (isIn) {
-                    PointerEventType.Enter
-                } else {
-                    PointerEventType.Exit
+                event.type = when {
```
Comment (olonho) [inline compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/input/pointer/HitPathTracker.kt:394]: Rather hard to read logic.
Maybe smth like
```
if (event.type == PointerEventType.Move) {
    event.type = when {
       !wasIn && isIn && !hasEntered -> PointerEventType.Enter
       wasIn && !isIn && hasEntered -> PointerEventType.Exit
       else -> PointerEventType.Move
   }
}
if (event.type == PointerEventType.Enter) hasEntered  = true
if (event.type == PointerEventType.Exit) hasEntered  = false
```

Comment (igordmn) [reply compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/input/pointer/HitPathTracker.kt:394]: Thanks, looks better.

But I can't remove `event.type == PointerEventType.Exit`. The logic is asymmetric, unfortunately.

Comment (olonho) [reply compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/input/pointer/HitPathTracker.kt:394]: Why? Cannot we force enter as well via keyboard navigation?

Comment (igordmn) [reply compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/input/pointer/HitPathTracker.kt:394]: > force enter as well via keyboard navigation?

Enter is pure pointer/mouse thing.

But I realized there is no special case, I just mask a bug from the desktop code, where we don't properly check the owner's bounds.

Fixed it in https://github.com/JetBrains/androidx/pull/109, and refactored this PR.

**Review by olonho — APPROVED** [review]

**Pull request merged**